### PR TITLE
Df vtk reader

### PR DESCRIPTION
**Pull request by dflanderka**

### CODE REVIEW

File: src/mesh/msh_vtkreader.cc [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+
+/*******************************************************************
+ * Helper methods
+ */
+template<typename T>
+T read_binary_value(std::istream &data_stream)
+{
+	T val;
+	data_stream.read(reinterpret_cast<char *>(&val), sizeof(val));
```
Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: Use `get` method instead of `read`.

File: src/mesh/msh_vtkreader.cc [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+		Tokenizer tok(file_name);
+		if (! tok.skip_to("AppendedData"))
+			THROW(GmshMeshReader::ExcMissingSection() << GmshMeshReader::EI_Section("AppendedData") << GmshMeshReader::EI_GMSHFile(tok.f_name()) );
+		else {
+			appended_pos_ = tok.get_position().file_position_;
+		}
+	}
+
+	char c;
+	appended_stream_->seekg(appended_pos_);
```
Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: Close tokenizer and stream, reopen in binary mode.

File: src/mesh/msh_vtkreader.cc (hunk 1 above)
Comment (dflanderka) [inline src/mesh/msh_vtkreader.cc]: We need use read method for reading binary data. Method get is using for reading strings ending with new-line character (see http://www.cplusplus.com/reference/istream/istream/get/).

Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: You are right.

File: src/mesh/msh_gmshreader.cc [hunk 3]
```diff
@@ -291,48 +289,44 @@ void GmshMeshReader::read_data_header(GMSH_DataHeader &head) {
...
+typename ElementDataCache<T>::ComponentDataPtr GmshMeshReader::get_element_data( std::string field_name, double time,
+		unsigned int n_entities, unsigned int n_components, bool &actual, std::vector<int> const & el_ids, unsigned int component_idx)
 {
     using namespace boost;
+    //
 
-    GMSH_DataHeader actual_header = find_header(search_header.time, search_header.field_name);
-    if ( !current_cache_->is_actual(actual_header.time, search_header.field_name) ) {
+    GMSH_DataHeader actual_header = find_header(time, field_name);
+    if ( !current_cache_->is_actual(actual_header.time, field_name) ) {
```
Comment (jbrezmorf) [inline src/mesh/msh_gmshreader.cc]: Only the base class will have templated get_element_data method, this method creates CacheData<T>  
entry and pass it as CacheDataBase (non templated , similar to OutputDataBase) to internal reader method
read_data. This method is virtual with separate implementation for GMSH and VTK (and separate from ascii and binary). These methods call virtual methods of CacheData<T> to read individual data fields from ascii or binary streams.
1. move check of actual current_cache_ into base class
2. In base class create new ElementDataCache<T>

File: src/mesh/msh_vtkreader.cc [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+        ASSERT_EQ(3, point_header.n_components).error();
+        node_ids.resize(point_header.n_entities);
+        for (unsigned int i=0; i<point_header.n_entities; ++i) {
+        	el_ids.push_back(i);
+        	vtk_to_gmsh_element_map_.push_back(i);
+        }
+
+        typename ElementDataCache<double>::CacheData data_cache
+            = ElementDataCache<double>::create_data_cache(1, point_header.n_components*point_header.n_entities);
+        tmp_cache = new ElementDataCache<double>(point_header.time, point_header.field_name, data_cache);
```
Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: Move constructor for CacheData into ElementDataCache.

File: src/mesh/msh_vtkreader.cc [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+
+        // check compatible nodes
+        this->read_element_data(*tmp_cache, point_header, 1, point_header.n_components, el_ids);
```
Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: rename tmp_cache to node_cache

File: src/mesh/msh_vtkreader.cc [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+        std::vector<double> &vec = *( data_cache[0] );
```
Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: node_vec

File: src/mesh/msh_vtkreader.cc [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+        ASSERT_EQ(vec.size(), point_header.n_components*point_header.n_entities).error();
+        for (unsigned int i=0; i<point_header.n_entities; ++i) {
+            Node node(vec[3*i], vec[3*i+1], vec[3*i+2]);
```
Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: Create arma::vec directly and pass to BIH::find_point

File: src/mesh/msh_vtkreader.cc [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+            int found_i_node = -1;
+            bih_tree.find_point(node.point(), searched_elements);
+
+            for (std::vector<unsigned int>::iterator it = searched_elements.begin(); it!=searched_elements.end(); it++) {
+                ElementFullIter ele = mesh.element( *it );
+                FOR_ELEMENT_NODES(ele, i_node)
+                {
+                    i_elm_node = mesh.node_vector.index(ele->node[i_node]);
+        			Node &elm_node = mesh.node_vector[ i_elm_node ];
+                    if (elm_node == node) {
```
Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: use some tolerance

File: src/mesh/msh_vtkreader.cc [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+        this->read_element_data(*tmp_cache, offset_header, 1, offset_header.n_components, el_ids);
+
+        offsets = *( data_cache[0] );
+        delete tmp_cache;
+    }
+
+    {   // read elements
+        MeshDataHeader con_header = this->find_header(0.0, "connectivity");
+        con_header.n_entities = offsets[offsets.size()-1];
+        for (unsigned int i=el_ids.size(); i<con_header.n_entities; ++i) {
```
Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: Why we need to init these arrays here.

File: src/mesh/msh_vtkreader.cc [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+
+        std::vector<unsigned int> &vec = *( data_cache[0] );
+        vector<unsigned int> node_list;
+        vector<unsigned int> result_list;
+        vtk_to_gmsh_element_map_.clear();
+        vtk_to_gmsh_element_map_.resize(offsets.size());
+        unsigned int i_con = 0; // iterate trough connectivity data
+        for (unsigned int i=0; i<offsets.size(); ++i) {
+            for ( ; i_con<offsets[i]; ++i_con ) {
+                node_list.push_back( node_ids[vec[i_con]] );
```
Comment (jbrezmorf) [inline src/mesh/msh_vtkreader.cc]: Better names for `vec`, `tmp_cache` ...

**Pull request merged**